### PR TITLE
Implementation of certificate revocation functionality

### PR DESCRIPTION
**Pull request by Maladie**
**Description**

Changes proposed in this pull request:

- Adds external and internal endpoints for revoking certificates

**Related issue(s)**

#2314

### CODE REVIEW

File: components/connector-service/internal/externalapi/revocationhandler.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import (
+	"github.com/kyma-project/kyma/components/connector-service/internal/apperrors"
+	"github.com/kyma-project/kyma/components/connector-service/internal/certificates"
+	"github.com/kyma-project/kyma/components/connector-service/internal/certificates/revocationlist"
+	"github.com/kyma-project/kyma/components/connector-service/internal/httphelpers"
+	"net/http"
+)
+
+const (
+	CertificateHeader = "Certificate-Header"
```
Comment (Szymongib) [inline components/connector-service/internal/externalapi/revocationhandler.go]: I suggest changing the header name to something like `Client-Certificate`

File: tests/connector-service-tests/test/testkit/securedconnectorclient.go [hunk 2]
```diff
@@ -72,6 +73,17 @@ func (cc securedConnectorClient) RenewCertificate(t *testing.T, url string, csr
 	return &certificateResponse, errorResp
 }
 
+func (cc securedConnectorClient) RevokeCertificate(t *testing.T, url string, csr string) *http.Response {
+	request, err := http.NewRequest(http.MethodPost, url, bytes.NewBuffer([]byte{}))
+	require.NoError(t, err)
+	request.Header.Add("Certificate-Header", csr)
```
Comment (Szymongib) [inline tests/connector-service-tests/test/testkit/securedconnectorclient.go]: As discussed, we should not set header here

File: tests/connector-service-tests/test/testkit/connectorclient.go [hunk 3]
```diff
@@ -126,8 +146,13 @@ func getRequestWithHeaders(t *testing.T, url string, headers map[string]string)
 func parseErrorResponse(t *testing.T, response *http.Response) *Error {
 	logResponse(t, response)
 	errorResponse := ErrorResponse{}
-	err := json.NewDecoder(response.Body).Decode(&errorResponse)
-	require.NoError(t, err)
+	decoder := json.NewDecoder(response.Body)
+
+	if decoder.More() {
+		err := decoder.Decode(&errorResponse)
```
Comment (Szymongib) [inline tests/connector-service-tests/test/testkit/connectorclient.go]: I do not think we need to check for more, as we do not expect multiple error responses.

File: tests/connector-service-tests/test/testkit/connectorclient.go [hunk 4]
```diff
@@ -66,6 +67,25 @@ func (cc connectorClient) CreateToken(t *testing.T) TokenResponse {
 	return tokenResponse
 }
 
+func (cc connectorClient) RevokeCertificate(t *testing.T, revocationUrl, csr string) *http.Response {
+	revocationBody := RevocationBody{
+		Hash: csr,
```
Comment (Szymongib) [inline tests/connector-service-tests/test/testkit/connectorclient.go]: As discussed we should pass here hash of the certificate instead of CSR.

Comment (franpog859) [reply tests/connector-service-tests/test/testkit/connectorclient.go]: Done!

File: tests/connector-service-tests/test/testkit/securedconnectorclient.go (hunk 2 above)
Comment (franpog859) [reply tests/connector-service-tests/test/testkit/securedconnectorclient.go]: Done!

File: components/connector-service/internal/certificates/middlewares/revocationcheck.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	return &revocationCheckMiddleware{
+		revocationList: revocationList,
+	}
+}
+
+func (rcm revocationCheckMiddleware) Middleware(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		certificate := r.Header.Get(externalapi.CertificateHeader)
+
+		hash, e := certificates.CalculateHash(certificate)
```
[the remark this answers is not in the document]
Comment (akgalwas) [reply components/connector-service/internal/certificates/middlewares/revocationcheck.go]: Done

File: components/connector-service/internal/certificates/middlewares/revocationcheck_test.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+		// given
+		testCert := "testCert"
+		hashedTestCert := "f21139ef2b82d02ee73a56c5c73c053fbafa3480a0b35459cba276b0667c57fc"
+		handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.WriteHeader(http.StatusOK)
+		})
+
+		req, err := http.NewRequest("POST", "/", nil)
+		req.Header.Set(externalapi.CertificateHeader, testCert)
+		require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (akgalwas) [reply components/connector-service/internal/certificates/middlewares/revocationcheck_test.go]: Done

File: components/connector-service/internal/certificates/middlewares/revocationcheck_test.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+		// when
+		resultHandler := middleware.Middleware(handler)
+		resultHandler.ServeHTTP(rr, req)
+
+		// then
+		assert.Equal(t, http.StatusInternalServerError, rr.Code)
+	})
+
+	t.Run("should return http code 500 when error occured during hash calculation", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (akgalwas) [reply components/connector-service/internal/certificates/middlewares/revocationcheck_test.go]: Done

File: components/connector-service/internal/externalapi/revocationhandler.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+	return &revocationHandler{
+		revocationList: revocationList,
+	}
+}
+
+func (handler revocationHandler) Revoke(w http.ResponseWriter, r *http.Request) {
+
+	hash, appError := handler.getCertificateHash(r)
+
+	if appError != nil {
```
[the remark this answers is not in the document]
Comment (akgalwas) [reply components/connector-service/internal/externalapi/revocationhandler.go]: Done

File: components/connector-service/internal/externalapi/revocationhandler.go (hunk 1 above)
Comment (akgalwas) [reply components/connector-service/internal/externalapi/revocationhandler.go]: Done

File: components/connector-service/internal/certificates/middlewares/revocationcheck_test.go [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+		// when
+		resultHandler := middleware.Middleware(handler)
+		resultHandler.ServeHTTP(rr, req)
+
+		// then
+		assert.Equal(t, http.StatusOK, rr.Code)
+	})
+
+	t.Run("should return http code 500 when error occured on contains method call", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (akgalwas) [reply components/connector-service/internal/certificates/middlewares/revocationcheck_test.go]: Done

File: components/connector-service/internal/certificates/middlewares/revocationcheck.go [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+		if e != nil {
+			httphelpers.RespondWithErrorAndLog(w, apperrors.Internal("Failed to calculate hash. Certificate could not be unescaped"))
+			return
+		}
+
+		contains, err := rcm.revocationList.Contains(hash)
+
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (akgalwas) [reply components/connector-service/internal/certificates/middlewares/revocationcheck.go]: Done

File: components/connector-service/README.md [hunk 11]
```diff
@@ -36,6 +36,7 @@ The Connector Service has the following parameters:
 - **runtimesInfoURL** - URL at which the management information for runtimes is available. If not provided, it bases on `connectorServiceHost`.
 - **certificateValidityTime** - Time until which the certificates that the service issues are valid. The default value is 90 days.
 - **central** - Determines whether the Connector Service works in the central mode.
+- **revocationConfigMapName** - Name of the config map containing revoked certificates list.
```
Comment (kazydek) [inline components/connector-service/README.md]:
```suggestion
- **revocationConfigMapName** - Name of the ConfigMap containing the revoked certificates list.
```

Comment (akgalwas) [reply components/connector-service/README.md]: Done.

**Pull request merged**